### DISCUSSION TITLE
Small things

### DISCUSSION THREAD

**Discussion by gboudreau** [econumo/discussions#1] [General]
1. On this page: https://econumo.com/docs/self-hosting/multi-currency/
The link for the repo is the same as the link to the releases.

2. In the sample .env @ https://github.com/econumo/econumo-ce/blob/main/.env.example
Having the `OPEN_EXCHANGE_RATES_*` values makes me think the docker container will use those values, but it won't, right? I need to set up the `currency-loader` separately?  
Maybe remove those values from .env.example in that repo; they are in the file found in the `currency-loader` repo.

3. It would be easier to set up the `currency-loader` if it was a docker image too.

4. The readme of `currency-loader` is sparse on details on how and when it should run, the first time or regularly (daily I guess?) My guess is that it should be run without arguments first (or when I add currencies), and then, every day, run it with `--date` with yesterday's date?

5. When I try to generate a new JWT key by running `bin/console lexik:jwt:generate-keypair --overwrite`, I get an error:
```
[2024-12-27T20:37:18.519508+00:00] console.DEBUG: Command "lexik:jwt:generate-keypair --overwrite" exited with code "1" {"command":"lexik:jwt:generate-keypair --overwrite","code":1} []

In Filesystem.php line 305:
                                                                                                                                                                  
  Cannot rename "/tmp/private.pemcfnfhC" to "/var/www/config/jwt/private.pem": rename(/tmp/private.pemcfnfhC,/var/www/config/jwt/private.pem): Permission denied  
```

Checking the automatically-created `jwt` folder, I see it is owner by `root:root`:
```
$ ls -l
drwxr-xr-x 2   82   82 10 Dec 27 15:27 db
drwxr-xr-x 2 root root 10 Dec 27 15:39 jwt
```
But if I wait 30-40s after the container was started, then the folder's ownership changes, and then it works as expected.  
It seems to take more than 1 minute on my system, between the time I start the container, and the time `docker-composer logs` shows anything.  
Maybe mention that, in your instructions about running this command.  
And/or maybe you could output (in docker logs) as soon as the container starts, asking the user to wait until one sees `nginx entered RUNNING state`

6. ~How can I find the value to use as `ECONUMO_SYSTEM_API_KEY`, in the `currency-loader` config? I didn't find anything about this [here](https://econumo.com/docs/api/) or in the Swagger doc.~ *update: I saw, looking in the backend code, that I need to define `ECONUMO_SYSTEM_API_KEY` in both the econumo container and in `currency-loader` to the same value.
 
7. In the onboarding guide, the `Connect with your partner` section should probably be hidden, for self-hosted instances, right? It says `please visit [Settings](https://econumo.home.patati.ca/#/settings) -> .`, which I guess is broken because the page it is trying to direct me too doesn't exist in the self-hosted version.

8. Wouldn't it make more sense for the `currency-loader` to be coded in PHP in the backend, instead of using a separate tool? Just ask for the open exchange rates config values (API key, currencies, base) in the .env of the docker container, and document how we should set up a cron job to call a specific URL, which would use those values to fetch from OER, and insert directly in the DB without having to use the API.

9. When documenting what we need to put in `OPEN_EXCHANGE_RATES_TOKEN`, you say `your Open Exchange Rates API key`, but OER never calls it that. It's called an App ID on their side of things.

10. Your `docker-compose` indicates we should mount a db volume in `/var/www/db`, but the DB is in fact in `/var/www/var/db`.

11. `currency-loader` can send the currencies I defined to econumo, but it fails to load rates:
```
./currency-loader
Currencies sent to econumo API successfully: {"items":["CAD","USD"]}
2024/12/27 16:45:27 Failed to fetch currency rates. Status code: 403
```
If I try the URL manually, I can see why:
```
$ curl 'https://openexchangerates.org/api/latest.json?app_id=MY_APP_ID_HERE&base=CAD'
{
  "error": true,
  "status": 403,
  "message": "not_allowed",
  "description": "Changing the API `base` currency is available for Developer, Enterprise and Unlimited plan clients. Please upgrade, or contact support@openexchangerates.org with any questions."
}
```
120 USD a year is way more than I am willing to pay for this. I'll just use `ECONUMO_CURRENCY_BASE=USD` in `currency-loader`, and change the main.go code to create a JSON sent to the above endpoint with `rate = 1/rate`.

12. For `POST /system/import-currency-rates`, the server returns a 500 error, when trying to save values that have already been sent:
```
$ ./currency-loader --date=2024-12-21
Currencies sent to econumo API successfully: {"items":["USD","CAD"]}
Currency rates sent to econumo API successfully: {"timestamp":1734825577,"base":"CAD","items":[{"code":"USD","rate":"0.695638"},{"code":"CAD","rate":"1.000000"}]}

$ ./currency-loader --date=2024-12-21
Currencies sent to econumo API successfully: {"items":["CAD","USD"]}
2024/12/27 17:43:57 Failed to send data to econumo API. Status code: 500
```
I think it would be better to allow this endpoint to overwrite an existing value. Maybe it does, and this error just occurs because I am sending exactly the same value..?

13. When the server returns a 500 error, there is no log I can look into to understand what is happening. I would guess that this is because I have `APP_ENV=prod` in my `.env`, as suggested? Maybe document how to see full stack for exceptions/error, in your `.env.example`?

14. When I tried to change the base currency:
```
$ docker exec -it -u www-data econumo sh -c 'echo $ECONUMO_CURRENCY_BASE'
CAD
```
I then had to run `currency-loader`, then logout and re-login for the UI to allow me to choose CAD in the settings.

More later, as I find other things.

**Comment by dmtrkzntsv** [comment]
I've fixed #10. I somehow missed that.

Thank you for the detailed feedback! I will review, validate and see what I can improve in the next couple of days

**Comment by dmtrkzntsv** [comment]
### Installation (1, 2, 5, 10, 13)
In the repository at https://github.com/econumo/econumo-ce
- I updated the README file to include debugging information (a new page is available at https://econumo.com/docs/self-hosting/debug/  for reference).
- I removed currency-related variables from `.env.example`, as you suggested.
- Added an example of the health-check.

### API details 
I updated the API documentation at https://econumo.com/docs/api/
- To clarify the difference between the System API and the User API.
- To include information about health checks.

### Onboarding bug (7)
It is a bug, I will fix it in the next release.

### Multi-currency
I updated the documentation at https://econumo.com/docs/self-hosting/multi-currency/ and made significant changes to the README in the repository at https://github.com/econumo/currency-loader, trying to answer your questions. Please, have a look.

Answering your other questions:

In short, you don't need to use `CAD` as your base currency; `USD` is perfectly fine. I've got 90% of my accounts in `CAD`, and I use `USD` as my base. It's important to note that the base currency is not the same as your default currency in the app; the base currency is only used for exchange conversions.

(UPDATE) I now see a 500 response when updating the currency list. I will fix it!

Regarding the creation of a separate script to update it:
I initially had it as a command, but I removed it from the project for two reasons:
1. I don't want Econumo to depend on any optional third-party vendor.
2. At one point, I was using DigitalOcean App Platform, which has no way to run a command on a schedule.

These commits provide more context:
- https://github.com/econumo/econumo-backend/commit/6ee7ef6831c4f9085765fd91ac62aa57274f29b0
- https://github.com/econumo/econumo-backend/commit/0f98297d8ba58d7b20ab6ae372fd177580289b10

By exposing an API, you avoid dependency on infrastructure limitations or vendors. I used GitHub Actions (https://github.com/econumo/currency-loader/blob/master/.github/workflows/daily-update.yml) to update my Econumo instance for a while, but later I switched to a different hosting service and began using cron jobs.

@gboudreau I hope I answered all questions! Thank you for your contribution! Let me know If I can fix/improve/clarify something else

**Reply by dmtrkzntsv** [reply]
Additionally, Econumo supports the following customization mechanisms if needed:

- [https://econumo.com/docs/self-hosting/tweaking-frontend/](https://econumo.com/docs/self-hosting/tweaking-frontend/)
- [https://econumo.com/docs/self-hosting/tweaking-backend/](https://econumo.com/docs/self-hosting/tweaking-backend/)

**Comment by gboudreau** [comment]
15. During onboarding, when I added a first account, I received an error in the (browser) console. The account was added, and the `Add account` step of the onboarding is checked.
<img width="1365" alt="Screenshot 2024-12-29 at 15 21 39" src="https://github.com/user-attachments/assets/d12dc2df-1ae7-42e2-8c16-ad27048c5ba8" />
<img width="514" alt="image" src="https://github.com/user-attachments/assets/cf3bb6ce-5cab-4fb8-bb3f-7de53fa0a877" />

16. In the date selector, if you click the selected date (ie. click the 29 round button), the date changes to 1900-01-01:
<img width="304" alt="image" src="https://github.com/user-attachments/assets/2eedcacc-5319-4c2e-bb86-955659967500" />
<img width="298" alt="image" src="https://github.com/user-attachments/assets/c84fe032-8656-4697-991d-e22d1fc728fb" />

17. There seems to be margin missing between the account and amount, in Expense and Income tabs:
<img width="508" alt="image" src="https://github.com/user-attachments/assets/cb0dfa21-10dd-4ab8-999d-45716a0da016" />

18. When you click the button to add a Transaction, the focus is moved automatically to the amount field, which I guess makes for faster/easier input, but having the field focused removes the placeholder text. When clicking that button the first time, we thus never see the placeholder text, and I was wondering what I should enter in the focused field. Maybe it would be better to simply keep the placeholder text visible until the field is not empty (i.e. continue showing it even when the field is selected)?
<img width="499" alt="image" src="https://github.com/user-attachments/assets/62de06c9-8e21-4416-b5b3-e169c5f4a688" />

19. It would be nice to be able to select multiple tags for a single transaction. For example, I use tags for my children, and some expenses I tag with 1+ children who "participated". eg if I go to the restaurant with 2 kids, I'd tag that txn with myself + those 2 kids. Same for family outings (cinema, ski, etc.)

20. I see that your categories are per txn type (expense vs income categories are different). Maybe you should show this "type", in the list of categories in the settings page? Either using tabs, to separate them in two lists, or by using a Income or Expense label (before or after the category name)?

21. I often (in my own system) use the same category for both expenses and income. For example, if I do a music gig, and get paid for the whole band (eg. 1000$), then I'll e-transfer each band member's pay to them (eg 2x 300$), so I categorize all those transactions together, for the sum (total paid minus all e-transfers) (eg. 400$) to appear as what I was paid for this gig. How can I track those transactions, to appear as a 400$ income, in the budget view?

22. I tested two expense transactions in the same category, but with different tags. But I see those together in the same tag, in the budget report:
<img width="1532" alt="image" src="https://github.com/user-attachments/assets/e7b4cca9-8b85-4d11-a357-4462bfc2958e" />
<img width="1499" alt="image" src="https://github.com/user-attachments/assets/8ac534b8-70a0-498c-b624-ac8b144e8a1c" />

**Reply by dmtrkzntsv** [reply]
15. Noted

16. Noted. I'm not satisfied with Quasar's default calendar widget, so I plan to replace it with another option in future releases.

17. It's intentional! :)

18. Agreed! It used to work that way, but I lost that feature with the introduction of the calculator widget. I've added a fix, and it will be included in v0.5.1.

19. The current concept is different; it uses a tag to group various spendings. It also organizes these spendings in the budget module under that tag. Having multiple tags would disrupt this behaviour.
I need to gather more use cases on this. How do you use these tags, and what value do they provide for you?

20. I agree. I've added it to my backlog

21. Currently, the only option is to change the original amount from 1000 to 400. However, I have a concept for "returns" in the backlog. Essentially, you would create a follow-up income "return" transaction, and as a result that would reflect 400 in the budget.

22. It is a bug, but I’ve already fixed it and will include the update in v0.5.1.
